### PR TITLE
Update forest.hpp for v3

### PR DESCRIPTION
**Pull request by james-d-mitchell**

### CODE REVIEW

File: include/libsemigroups/forest.hpp [hunk 1]
```diff
@@ -93,59 +89,49 @@ namespace libsemigroups {
...
-            _edge_label.cend(), n, static_cast<size_t>(UNDEFINED));
-        _parent.insert(_parent.cend(), n, static_cast<size_t>(UNDEFINED));
-      } catch (...) {
-        _edge_label.resize(old_nr_nodes);
-        _parent.resize(old_nr_nodes);
-        throw;
-      }
+    //! \param that the Forest for comparison.
+    //!
+    //! \returns whether or not the Forest objects are not equal.
```
Comment (Joseph-Edwards) [inline include/libsemigroups/forest.hpp:109]: Underneath `\brief Compare Forest objects for inequality`, should we have something like 
```
This function returns the negation of `Operator==(that)`
```

The fact that `operator!=` jumps straight into the parameters whereas `operator==` doesn't feels a bit inconsistent. Not sure about this one though

File: include/libsemigroups/forest.hpp [hunk 2]
```diff
@@ -216,15 +226,26 @@ namespace libsemigroups {
...
+    //!
+    //! \param i the node whose parent is sought.
+    //!
+    //! \returns
+    //! A \ref node_type.
+    //!
+    //! \complexity
+    //! Constant
+    //!
+    //! \warning No checks are performed on the arguments of this function.
```
Comment (Joseph-Edwards) [inline include/libsemigroups/forest.hpp:243]: Same as with `operator!=`. The omission of a description after the brief description doesn't look quite right to me.

File: include/libsemigroups/forest.hpp [hunk 3]
```diff
@@ -237,135 +258,160 @@ namespace libsemigroups {
...
-      }
-    }
+    //! \brief Modifies \p w to contain the labels of the edges on the path
+    //! from a root node to \p i.
+    //!
+    //! This function modifies its first argument \p w in-place to contain the
+    //! labels of the edges on the path from a root node to node \p i.
+    //!
+    //! \param w value to contain the result
+    //! \param i the node
```
Comment (Joseph-Edwards) [inline include/libsemigroups/forest.hpp:324]:
```suggestion
    //! \param w value to contain the result.
    //! \param i the node.
```

File: include/libsemigroups/forest.hpp [hunk 4]
```diff
@@ -237,135 +258,160 @@ namespace libsemigroups {
...
+    void path_to_root_no_checks(word_type& w, node_type i) const;
 
-    // TODO doc
+    //! \brief Returns a word containing the labels of the edges on the path
+    //! from a root node to \p i.
+    //!
+    //! This function returns a word containing the labels of the edges on the
+    //! path from a root node to node \p i.
+    //!
+    //! \param i the node
```
Comment (Joseph-Edwards) [inline include/libsemigroups/forest.hpp:335]:
```suggestion
    //! \param i the node.
```

File: include/libsemigroups/forest.hpp [hunk 5]
```diff
@@ -237,135 +258,160 @@ namespace libsemigroups {
...
 
-    // TODO doc
+    //! \brief Modifies \p w to contain the labels of the edges on the path
+    //! from a root node to \p i.
+    //!
+    //! This function modifies its first argument \p w in-place to contain the
+    //! labels of the edges on the path from a root node to node \p i.
+    //!
+    //! \param w value to contain the result
+    //! \param i the node
```
Comment (Joseph-Edwards) [inline include/libsemigroups/forest.hpp:351]:
```suggestion
    //! \param w value to contain the result.
    //! \param i the node.
```

File: include/libsemigroups/forest.hpp [hunk 6]
```diff
@@ -237,135 +258,160 @@ namespace libsemigroups {
...
     }
 
-    // TODO doc
+    //! \brief Returns a word containing the labels of the edges on the path
+    //! from a root node to \p i.
+    //!
+    //! This function returns a word containing the labels of the edges on the
+    //! path from a root node to node \p i.
+    //!
+    //! \param i the node
```
Comment (Joseph-Edwards) [inline include/libsemigroups/forest.hpp:366]:
```suggestion
    //! \param i the node.
```

File: src/forest.cpp [hunk 7]
```diff
@@ -24,17 +24,92 @@
...
+//   rx::seq<Forest::node_type>() | rx::transform([&f](auto i) {
+//     word_type w;
+//     while (f.parent_no_checks(i) != UNDEFINED) {
+//       w.push_back(f.label_no_checks(i));
+//       i = f.parent_no_checks(i);
+//     }
+//     return w;
+//   });
+// }
+//}  // namespace forest
```
Comment (Joseph-Edwards) [inline src/forest.cpp:66]: Is this commented out deliberately?

**Review by Joseph-Edwards — CHANGES_REQUESTED** [review]
Other than the omission of some descriptions after the `\brief` descriptions (which I'm not convinced is a bad thing, but probably worth bringing up), and the other minor comments bellow, this looks good to me :)

File: include/libsemigroups/dot.hpp [hunk 8]
```diff
@@ -599,9 +599,16 @@ namespace libsemigroups {
...
+
+  // TODO(doc)
   std::string to_human_readable_repr(Dot::Node const& n);
+
+  // TODO(doc)
   std::string to_human_readable_repr(Dot::Edge const& e);
+
+  // TODO(doc)
   std::string to_human_readable_repr(Dot::Kind const&   k,
                                      std::string const& sep = "::");
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply include/libsemigroups/dot.hpp:613]: I just noticed when working on this PR that there wasn't any doc for these 4 functions, and added the TODOs. So probably for before v3 but not in this PR

File: src/forest.cpp (hunk 7 above)
Comment (james-d-mitchell) [reply src/forest.cpp:66]: Yes, it seems not to be used anywhere and there's the TODO about improving it.

**Comment by james-d-mitchell** [comment]
@Joseph-Edwards thanks for the review, I've updated it according to your comments. I think I expect the pybind11 to fail.

**Comment by Joseph-Edwards** [comment]
> @Joseph-Edwards thanks for the review, I've updated it according to your comments. I think I expect the pybind11 to fail.

Should the pybind11 be fixed in this PR, or is that good to merge now?

**Comment by james-d-mitchell** [comment]
@Joseph-Edwards merge away, it's chicken and egg with this PR and the libsemigroups_pybind11 one

**Review by Joseph-Edwards — APPROVED** [review]

**Pull request merged**